### PR TITLE
Constant Range Support

### PR DESCRIPTION
**Pull request by zhengyangl**

### CODE REVIEW

**Comment by regehr** [comment]
I don't understand the parts of this patch that are doing stuff with demanded bits, why is that in here?

**Comment by zhengyangl** [comment]
@regehr Before this PR, all the constants we're synthesizing stay in the RHSs. But in constant range synthesis, we now have reserved constants in LHSs: For query [(x<= LHS') && (x<LHS'+c) == True], here, (x<= LHS) && (x< LHS+c) is the LHS and True is the RHS. Since the reserved constant now occurs in LHS, after the first query, we need to call getInstCopy() on LHS to fill it with the model returned from the solver. This triggers the bug where the demandedbits are not initialized properly in function getPhi() and getInstCopy(), which leads to wrong synthesis results.

**Comment by zhengyangl** [comment]
Before this pull request, the model filling only works on RHS, and the bug is hidden because we're not having phi's synthesized.

**Comment by zhengyangl** [comment]
Also I've thought about reverse the query, like [True == (X<= LHS') && (X<LHS'+c)], where True as LHS and (x<= LHS') && (x<LHS'+c). By setting the query this way we can eliminate the extra getInstCopy() call on LHS. However, the refinement check are needed here and the direction matters. The reversed query can lead to over-conservative results.

File: lib/Extractor/Solver.cpp [hunk 1]
```diff
@@ -472,6 +472,94 @@ class BaseSolver : public Solver {
     return !IsSat;
   }
 
+  std::error_code testRange(const BlockPCs &BPCs,
+                            const std::vector<InstMapping> &PCs,
+                            Inst *LHS, llvm::APInt &C,
+                            llvm::APInt &X,
```
Comment (regehr) [inline lib/Extractor/Solver.cpp]: can you find a better name for X and SynthesisX?

Comment (zhengyangl) [reply lib/Extractor/Solver.cpp]: How about naming X as "Result" and naming SynthesisX as "ReservedResult"?

File: lib/Extractor/Solver.cpp [hunk 2]
```diff
@@ -472,6 +472,94 @@ class BaseSolver : public Solver {
...
+      IsFound = false;
+    } else {
+      IsFound = true;
+      X = ResultMap[SynthesisX];
+    }
+
+    return EC;
+  }
+
+  std::error_code range(const BlockPCs &BPCs,
```
Comment (regehr) [inline lib/Extractor/Solver.cpp]: can you make this function have a better name too?

File: lib/Extractor/Solver.cpp [hunk 3]
```diff
@@ -472,6 +472,94 @@ class BaseSolver : public Solver {
...
+    APInt BinSearchResultX, BinSearchResultC;
+    bool BinSearchHasResult = false;
+
+    while (L.ule(R)) {
+      APInt M = L + ((R - L)).lshr(1);
+      APInt BinSearchX;
+      bool Found = false;
+      EC = testRange(BPCs, PCs, LHS, M, BinSearchX, Found, IC);
+      if (EC) {
+        llvm::report_fatal_error("stopping due to error");
```
Comment (regehr) [inline lib/Extractor/Solver.cpp]: does this ever happen in practice?

File: lib/Extractor/Solver.cpp [hunk 4]
```diff
@@ -472,6 +472,94 @@ class BaseSolver : public Solver {
...
+      }
+      if (Found) {
+        R = M - 1;
+
+        // record result
+        BinSearchResultX = BinSearchX;
+        BinSearchResultC = M;
+        BinSearchHasResult = true;
+      } else {
+        if (L == R) break;
```
Comment (regehr) [inline lib/Extractor/Solver.cpp]: put break on a new line please

**Comment by regehr** [comment]
looks great except for a few minor changes

File: lib/Extractor/Solver.cpp (hunk 3 above)
Comment (zhengyangl) [reply lib/Extractor/Solver.cpp]: Sorry, it wont happen, after calling testRange, either we found the range or we don't. I'll remove this line.

File: lib/Extractor/Solver.cpp (hunk 1 above)
Comment (regehr) [reply lib/Extractor/Solver.cpp]: sounds good

**Comment by zhengyangl** [comment]
Fixed the issues, please review @regehr

**Comment by regehr** [comment]
LGMT, thanks!

**Pull request merged**